src/main/java/org/jetbrains/bsp/bazel/server/DiagnosticsDispatcher.java

```diff
@@ -0,0 +1,139 @@
+package org.jetbrains.bsp.bazel.server;
+
+import ch.epfl.scala.bsp4j.BuildClient;
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier;
+import ch.epfl.scala.bsp4j.Diagnostic;
+import ch.epfl.scala.bsp4j.DiagnosticSeverity;
+import ch.epfl.scala.bsp4j.Position;
+import ch.epfl.scala.bsp4j.PublishDiagnosticsParams;
+import ch.epfl.scala.bsp4j.Range;
+import ch.epfl.scala.bsp4j.SourceItem;
+import ch.epfl.scala.bsp4j.TextDocumentIdentifier;
+import com.google.common.collect.Lists;
+import io.bazel.rules_scala.diagnostics.Diagnostics;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.bsp.bazel.common.Uri;
+
+public class DiagnosticsDispatcher {
+
+  private static final Logger LOGGER = LogManager.getLogger(DiagnosticsDispatcher.class);
+
+  private final BazelBspServer bspServer;
+  private final BuildClient bspClient;
+
+  public DiagnosticsDispatcher(BazelBspServer bspServer, BuildClient bspClient) {
+    this.bspServer = bspServer;
+    this.bspClient = bspClient;
+  }
+
+  public void collectDiagnostics(
+      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics,
+      BuildTargetIdentifier target,
+      String diagnosticsLocation)
+      throws IOException {
+    Diagnostics.TargetDiagnostics targetDiagnostics =
+        Diagnostics.TargetDiagnostics.parseFrom(Files.readAllBytes(Paths.get(diagnosticsLocation)));
+
+    for (Diagnostics.FileDiagnostics fileDiagnostics : targetDiagnostics.getDiagnosticsList()) {
+      LOGGER.info("Inserting diagnostics for path: " + fileDiagnostics.getPath());
+
+      filesToDiagnostics.put(
+          getUriForPath(fileDiagnostics.getPath()), convertDiagnostics(target, fileDiagnostics));
+    }
+  }
+
+  public void emitDiagnostics(
+      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics, BuildTargetIdentifier target) {
+    for (SourceItem source : bspServer.getCachedBuildTargetSources(target)) {
```

Comment:
nested fors can be extracted

Reply:
I agree but here I was not sure about the logic (and naming). What exactly does the code inside _if (bspClient != null) {_ do?

Reply:
In my opinion this predicate is always true

Reply:
I think it's unnecessary too, but I didn't want to break it. Still, I mean this:

`
for (List<PublishDiagnosticsParams> values : filesToDiagnostics.values()) {
          for (PublishDiagnosticsParams param : values) {
            bspClient.onBuildPublishDiagnostics(param);
          }
        }
`

What exactly is happening here and why do we need these three nested loops?

Reply:
> In my opinion this predicate is always true
I'm not sure how that has changed but since the BSP server needed access to the BEP client and the BEP client to the BSP client  (provided by the BSP server), there was at that point a race condition where an event could have been emitted before BEP server had access to it.

> What exactly is happening here and why do we need these three nested loops?
This is a bit of a more messy part of the code since this is to ensure backwards compatibility with older versions of bazel. Basically here, what happens is that diagnostics events coming from the BEP, after being converted to BSP diagnostics, have to be published to the client. It's three nested loops due to this: 
- a target has multiple diagnostics (usually one per source file)
- a  source file might have multiple diagnostics
- a BEP diagnostic event might have more than one BSP diagnostic

This can be flattened, but, other than, not sure if there are any improvements that can be done.

Reply:
> I'm not sure how that has changed but since the BSP server needed access to the BEP client and the BEP client to the BSP client (provided by the BSP server), there was at that point a race condition where an event could have been emitted before BEP server had access to it.

There is indeed something like this in _Server.java_ but there we have:

``
BepServer bepServer = new BepServer(bspServer, launcher.getRemoteProxy(), buildClientLogger);
bspServer.setBepServer(bepServer);
`` 
(where ``launcher.getRemoteProxy()`` is ``BuildClient bspClient``)

so I think it's not a problem, maybe only if this _launcher.getRemoteProxy()_ works in some strange way. What do you think?

> This is a bit of a more messy part of the code since this is to ensure backwards compatibility with older versions of bazel. Basically here, what happens is that diagnostics events coming from the BEP, after being converted to BSP diagnostics, have to be published to the client. It's three nested loops due to this:
> 
> * a target has multiple diagnostics (usually one per source file)
> * a  source file might have multiple diagnostics
> * a BEP diagnostic event might have more than one BSP diagnostic
> 
> This can be flattened, but, other than, not sure if there are any improvements that can be done.

Thanks for explaining! But still what we do here is we put things into _filesToDiagnostics_, then do these two loops over _filesToDiagnostics.values()_ to publish and then go to the next source without removing anything from _filesToDiagnostics_. So aren't some parts of _filesToDiagnostics_ published multiple times? Can't these two loops be moved out and after the one over the sources or am I misunderstanding this?

Reply:
I believe that the `filesToDiagnostics` has their values removed once another event for that target is received. We can't just remove them because a specific target might be cached and we still need to resend all diagnostics when a new compilation happens.
This leads to the same diagnostic having the possibility of being sent more than once during a single compilation yes (which is inefficient but is assumed as a possibility in the protocol), but your change would make us not be protocol-compliant. I believe there are ways of making this more efficient but they are more complicated and probably should not be addressed in this PR. Wdyt?

Reply:
Yup, I agree. Just wanted to understand what was going on there because it didn't make sense at first sight.